src/thunderbots/software/geom/util.cpp

```diff
@@ -204,13 +203,18 @@ bool intersects(const Circle &first, const Circle &second)
 
 bool intersects(const Ray &first, const Segment &second)
 {
-    auto isect = lineIntersection(first.getRayStart(), first.getDirection(),
-                                  second.getSegStart(), second.getEnd());
+    auto isect =
+        lineIntersection(first.getRayStart(), first.getRayStart() + first.getDirection(),
+                         second.getSegStart(), second.getEnd());
+    // If the infinitely long vectors defined by ray and segment intersect, check that the
+    // intersection is within their definitions
     if (isect.has_value())
     {
         return contains(first, isect.value()) && contains(second, isect.value());
     }
-    return collinear(first.getRayStart(), first.getDirection(), second.getSegStart());
+    // If there is no intersection, the ray and segment may be parallel, check if they are
+    // overlapped
+    return contains(second, first.getRayStart());
```

Comment:
Why did we make changes here? If it was because this function was doing something incorrect, could we add a unit test for the behavior that was incorrect to ensure we don't regress? (In general we should just try to make sure that changes to functions are accompanied by corresponding test(s) or test(s) changes)

Same goes for the changes to `vectorRectIntersect` below.

Reply:
The function was improperly using Rays. It treats the ray direction as a point along the ray's path, which is incorrect. A point along the ray's path will be ray.start() + n*ray.direction()

Reply:
Added unit tests that checks the case that ray.getDirection intersects, but not (ray.getStartPosition() + ray.getDirection())